casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -31,15 +28,59 @@ object Validate {
       if (isValid) {
         true.pure[F]
       } else {
-        Log[F].warn(ignore(b, "signature is invalid.")) *> false.pure[F]
+        for {
+          _ <- Log[F].warn(ignore(b, "signature is invalid."))
+        } yield false
       }
     } else {
-      Log[F].warn(
-        ignore(b, s"signature algorithm ${b.sigAlgorithm} is unsupported.")
-      ) *> false.pure[F]
+      for {
+        _ <- Log[F].warn(ignore(b, s"signature algorithm ${b.sigAlgorithm} is unsupported."))
+      } yield false
+    }
+
+  def blockSender[F[_]: Applicative: Log](b: BlockMessage,
+                                          genesis: BlockMessage,
+                                          dag: BlockDag): F[Boolean] =
+    if (b == genesis) {
+      true.pure[F] //genesis block has a valid sender
+    } else {
+      val weight = ProtoUtil.weightFromSender(b, dag.blockLookup)
+      if (weight > 0)
+        true.pure[F]
+      else
+        for {
+          _ <- Log[F].warn(
+                ignore(b, s"block creator ${PrettyPrinter.buildString(b.sender)} has 0 weight."))
+        } yield false
     }
 
-  def blockNumber[F[_]: Applicative: Log](b: BlockMessage, dag: BlockDag): F[Boolean] = {
+  def validateBlockSummary[F[_]: Monad: Log](block: BlockMessage,
+                                             genesis: BlockMessage,
+                                             dag: BlockDag): F[Either[BlockStatus, BlockStatus]] =
```

Comment:
I don't love the `Either[BlockStatus, BlockStatus]` return type. Just looking at that alone makes you wonder why use `Either` at all? I still think that taking `Valid` out of the `BlockStatus` trait and renaming the trait to `BlockError` would be better. The return type could then be `Either[BlockError, Valid]` (if you kept around the `Valid` object), or `Either[BlockError, BlockMessage]`. I prefer the latter, but both convey what is going on more clearly.